compiler/src/dotty/tools/dotc/transform/localopt/Devalify.scala

```diff
@@ -194,13 +198,22 @@ class Devalify extends Optimisation {
       } else if (t.symbol.is(CaseAccessor) && !t.symbol.is(Mutable))
         readingOnlyVals(rec)
       else false
+
     case Select(qual, _) if !t.symbol.is(Mutable) =>
-      readingOnlyVals(qual)
+      if (t.symbol == defn.SystemModule) {
+        // System.in is static final fields that, for legacy reasons, must be
+        // allowed to be changed by the methods System.setIn...
+        // https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5.4
+        false
+      } else
+        readingOnlyVals(qual)
+
     case t: Ident if !t.symbol.is(Mutable) && !t.symbol.is(Method) && !t.symbol.info.dealias.isInstanceOf[ExprType] =>
```

Comment:
Nothing says that there might not be an `Ident` corresponding to `System.in`, and in that case you would still inline it. Maybe add a `isReallyMutable` instead that you would use everywhere?